Added draft materials for the introductory module and the first data …

aliciacanipe opened the pull request:
…products module

These are the first draft of materials for the introduction module and the first data product module. The data product notebook will probably need to be reduced, but I wanted to start by providing too much information and then scaling it back. I have not added exercises just yet. I'll add them after I get some feedback.

camipacifici commented:
The presentation looks great!
My only suggestion would be to add your name (and maybe the names of the other people involved in this webinar?) to the title page. Your call anyway!
I will review the notebook now.

@bhilbert4 if you can, please review the notebook @aliciacanipe submitted. Thanks!

camipacifici commented:
The notebook also looks great!

Can you please add a requirements.txt file with the package versions you need to run the notebook? Something like the one here: https://github.com/spacetelescope/jwebbinar_prep/tree/test-env/test_environment
I will then run it in its correct environment.

This is the box folder where you can put the material:
https://stsci.box.com/s/hgu7drfwbjm982pgmlk55ruhl2gbwoyr

I would suggest creating a "plotting" function at the beginning so that you can keep the coding of the plots simple during the class.

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fits.info(uncal_obs)"
```
This duplicates the call to fits.info that was made just above the section 4 header.

aliciacanipe replied:
Agreed, I removed the extra fits.info

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "science_data = fits.getdata(uncal_obs,1)\n",
```
Maybe switch the 1 to 'SCI' here in order to be consistent with the 'with' statement below, and just to be more explicit?

aliciacanipe replied:
Didn't even know that was possible, I fixed it

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Headers containing information about the observation and the data parameters be accessed the standard way:"
```
"parameters can be"

aliciacanipe replied:
Fixed

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+    "\n",
+    "# or \n",
+    "\n",
+    "print('\\nNumber of data dimensions: ', science_headers['NAXIS'])\n",
```
If the goal is to start to introduce users to JWST jargon, maybe we should remove the NAXIS lines here and keep only the lines that use the primary header keywords?

aliciacanipe replied:
Good point, removed NAXIS reference

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the previous section, we mentioned [up-the-ramp sampling](https://jwst-docs.stsci.edu/understanding-exposure-times) for IR detectors. During an integration, the detectors accumulate charge while being read out multiple times non-destructively according to readout patterns. The readout process is non-destructive, leaving charge unaffected and in place (charge is not transferred between pix ...
```
while being read out multiple times non-destructively according to readout patterns -> while being read out multiple times non-destructively following predefined readout patterns
and then link to the JDox readout pattern page? 
https://jwst-docs.stsci.edu/near-infrared-camera/nircam-instrumentation/nircam-detector-overview/nircam-detector-readout-patterns

aliciacanipe replied:
Changed to "During an integration, the detectors accumulate charge while being read out multiple times following predefined readout patterns for the different instruments. The readout process is non-destructive, leaving charge unaffected and in place (charge is not transferred between pixels as in CCDs)."

bhilbert4 commented on pipeline_products_session/jwst-data-products-part1.ipynb:
```diff
@@ -0,0 +1,961 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### C.<font color='white'>-</font>Metadata<a class=\"anchor\" id=\"model-meta\"></a>"
```
It might be useful to have a cell that shows the result from <model>.meta.instance in order to see all the metadata entries at once.

aliciacanipe replied:
Thank god, I could not figure out what that attribute was. Added an example

bhilbert4 reviewed (approved):
This looks awesome. Very clear and easy to follow.

aliciacanipe commented:
Thanks @bhilbert4! I'll take a look and work on addressing your feedback

bhilbert4 commented on pipeline_products_session/jwst-data-products-part2.ipynb:
```diff
@@ -0,0 +1,1401 @@
+    "\n",
+    "## 6.1.<font color='white'>-</font>Imaging<a class=\"anchor\" id=\"s2-imaging\"></a>\n",
+    "\n",
+    "Stage 2 image processing applies additional instrumental corrections and calibrations that result in a fully calibrated individual exposure. Non-time series exposures use the CALWEBB_IMAGE2 configuration, which applies all applicable steps to the data. The CALWEBB_TSO-IMAGE2 configuration, on the other hand, should be used for time series exposures, for which some steps are set to be skipped ...
```
You might want to define "configuration" here, or perhaps re-word to indicate that there are different sets of pre-defined steps that are mode dependent.

aliciacanipe replied:
Fixed the statement

bhilbert4 commented on pipeline_products_session/jwst-data-products-part2.ipynb:
```diff
@@ -0,0 +1,1401 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 6.2.<font color='white'>-</font>Spectroscopy<a class=\"anchor\" id=\"s2-spectroscopy\"></a>\n",
```
You might want to define "configuration" here, or perhaps re-word to indicate that there are different sets of pre-defined steps that are mode dependent.

aliciacanipe replied:
Fixed

bhilbert4 commented on pipeline_products_session/jwst-data-products-part2.ipynb:
````diff
@@ -0,0 +1,1401 @@
+    "* **2D resampled image**\n",
+    "    * Data model: DrizProductModel\n",
+    "    * File suffix: ```_i2d```\n",
+    "    * Description: This is the output of the resample step and is only created for regular direct imaging observations (not for time series or coronagraphy 3D data sets). Note that this product is intended for quick-look use only and is not passed along as input to Stage 3 processing. Calibrated, but unrectified (```_cal```) products are used as input to Stage 3. \n",
````
It might be good to quickly define "resampled" here

aliciacanipe replied:
Fixed

bhilbert4 commented on pipeline_products_session/jwst-data-products-part2.ipynb:
```diff
@@ -0,0 +1,1401 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "detector_to_world = cal_image.meta.wcs.get_transform('detector', 'world')"
```
I think this whole WCS section is really useful. One small suggestion would be to show that the transformations can run on lists of pixels, i.e. it doesn't have to run on one pixel at a time.

aliciacanipe replied:
Good idea, added an example

bhilbert4 reviewed (approved):
Looks good! Not many suggestions. If you end up needing to save time, maybe you can skip over the description of the inputs and outputs for some of the modes (AMI, TSO, etc)?

bhilbert4 commented on pipeline_products_session/jwst-data-products-part3.ipynb:
```diff
@@ -0,0 +1,901 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "4.<font color='white'>-</font>Data products: stage 3 (combined, rectified exposures)<a class=\"anchor\" id=\"stage3\"></a>\n",
```
It might be helpful to define or describe "rectified" and "combined" here. e.g. "combined" = mosaicked.

aliciacanipe replied:
Fixed

cslocum commented:
I also am not able to access https://data.science.stsci.edu/redirect/JWebbinar/1_pipeline_products/jwebbinar_data/example_nircam_imaging_uncal.fits (needed in notebook part1).

aliciacanipe commented:
Hi all, I pushed updated versions of the notebooks based on feedback, including "live" (empty cells) and "static" (completed cells) versions. Exercises were added at the end of each notebook. The only thing that needs sorting is the data files. As @cslocum mentioned, I'm still having issues with the box redirect link. I wasn't able to get it working. I might need some technical help with that. @camipacifici FYI

camipacifici commented:
Looks great!! I really like the idea of having a live and a static version.

About the links, the folder is not in the DMD managed folder so we cannot use the human readable structure "https://data.science.stsci.edu/redirect/JWebbinar/1_pipeline_products/jwebbinar_data/". Please, use the non-human readable links for now. For example you would use this (https://stsci.box.com/s/mpbrc3lszdjif6kpcw1acol00e0mm2zh) for the example NIRCam uncal file. While, I actually do not see on box the "example_exercise_uncal.fits".

I will contact ITSD to see if we can create that redirect for our JWebbinar folder too.

camipacifici commented:
The other thing we need to do is to see how long it takes to run the "live" version. I can be your audience if you want to try to live code it.

bhilbert4 commented:
@camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks in on AWS perhaps everything will be much faster?

aliciacanipe commented:
Thanks @camipacifici, I'll try the ugly box links today and see if I can get the data to load, and also add the missing example exercise data. I'll go through and practice the live coding a couple times to see how long it takes, and reach out if I need a test audience member.

aliciacanipe commented:
Okay, I updated the simulations for the notebooks and fixed almost all of the Box links and downloads. I think these are my final updates. 

I was trying to get simulations for other instruments for the exercise in part2 but the NIRISS, NIRSPec, and MIRI LRS examples I found in the JDAT repo had not been through the right pipeline steps to be able to use in the exercise. So I just used another NRC image. Might be good to find another simulation. 

The other issue is that when you download a file from Box, you have to save it to your machine using FITS with a non-Box-junk filename so that the data models can read the file. However, I couldn't figure out how to do this with the source catalog (*cat.ecsv) because you can't use FITS, and Astropy.table wasn't working with Box. So right now the source catalog download won't work. I'll need some help there. @camipacifici

camipacifici commented:
Ah! I did not think about other types of files, sorry. How about having a zip file that gets downloaded and extracted locally?
Something like this:

boxlink = 'uglylink...blabla..."
boxfile = './blabla.zip'
urllib.request.urlretrieve(boxlink, boxfile)
zf = zipfile.ZipFile(boxfile, 'r')
zf.extractall()
observedfiles = './path to file/'
file1d = observedfiles + 'my_blabla_file.whatever'

This allows you to keep the names of the files and have just one ugly box link.
We might want to double check with @gsnyder206 if it is ok to do so on AWS.

And I see your new commit saying that you fixed the issue so maybe this is not relevant anymore, but could be useful for posterity. :)

aliciacanipe commented:
Sorry, now I've made my final final updates. @bhilbert4 provided a download function that works for the catalog so I switched it out for the part3 notebooks. They're good to go now, ready to try them out.

camipacifici commented:
If @bhilbert4 gives his thumb up, I will merge this PR!

camipacifici commented:
> @camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks on AWS perhaps everything will be much faster?

Making sure I do not miss this. @cslocum @gsnyder206 any ideas here?

bhilbert4 reviewed (approved):
Notebook 3 looks good to me! I ran it locally and everything worked well, and the content is good!

cslocum commented:
> @camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks on AWS perhaps everything will be much faster?

I wouldn't count on anything faster.  Once these notebooks are merged, we can test them out on the platform.

cslocum commented:
I will also mention that we likely cannot include the data files in the Docker image.  They are just too large (in fact, github will not even allow them to be committed).  We can either cache them in a central place, or have users download them in real-time (but this has cost implications).

I'll see if @gsnyder206 has any opinions here.  Also - Greg will be on vacation starting tomorrow, back next Tuesday, so I'm probably your best contact for this stuff.

camipacifici commented:
The pipeline webinars will need to use big files, either generated during the webinar or simply called form the notebook. Lets try and see what happens.
Merging this now!

cslocum commented:
Awesome!  I'll test things out!
